spread.yaml

```diff
@@ -0,0 +1,122 @@
+  # around. The purpose of this code is to fix some connectivity issues setup 
+  # spread functions to be more usable outside of the explicit task.yaml 
+  # sections.
+
+  # apt update is hanging on security.ubuntu.com with IPv6, prefer IPv4 over IPv6
+  cat <<EOF > gai.conf
+  precedence  ::1/128       50
+  precedence  ::/0          40
+  precedence  2002::/16     30
+  precedence ::/96          20
+  precedence ::ffff:0:0/96 100
+  EOF
+  if ! mv gai.conf /etc/gai.conf; then
+      echo "/etc/gai.conf is not writable, ubuntu-core system? apt update won't be affected in that case"
+      rm -f gai.conf
+  fi
+
+  # Take the MATCH and REBOOT functions from spread and allow our shell
+  # scripts to use them as shell commands. The replacements are real
+  # executables in tests/lib/bin (which is on PATH) but they source
+  # spread-funcs.sh written here, base on the definitions provided by SPREAD.
+  # This ensures that 1) spread functions define the code 2) both MATCH and
+  # REBOOT are executables and not functions, and can be called from any
+  # context.
+  type MATCH | tail -n +2 > "$TESTSLIB"/spread-funcs.sh
+  unset MATCH
+  type REBOOT | tail -n +2 >> "$TESTSLIB"/spread-funcs.sh
+  unset REBOOT
+
+  # prepare common uc20 image setup by repacking snaps, etc
+  "$TESTSLIB"/prepare-uc20.sh
+
+restore-each: |
+  # delete the nested VM image after each task finishes so we don't use too much
+  # disk space
+  . "$TESTSLIB/nested.sh"
+  cleanup_nested_core_vm
+
+# it takes a while to build everything while preparing the project
+warn-timeout: 20m
+# but it shouldn't take _too_ long...
+kill-timeout: 30m
+
+suites:
+  tests/spread/:
+    summary: Main spread tests
+    environment:
+      # since we are only testing uc20 here, disable kvm unless the env
+      # specifically knows kvm works and sets this to true
+      # kvm specifically does not work on GCE with focal L1 VM and focal (UC20)
+      # L2 VM's because the L2 VM intermittently and randomly hangs and GCE's
+      # hypervisor will restart the VM, there is not currently a known
+      # workaround
+      ENABLE_KVM: '$(HOST: echo "${SPREAD_ENABLE_KVM:-false}")'
+
+      # variants of various system setup below, turning on/off secure boot and
+      # enabling/disabling tpm, this ensures that all tests we write will be run
+      # against all variants
+
+      # TODO: are there other variants/knobs the initramfs we want to support
+      # testing?
+
+      # full encryption has TPM and secure boot
+      ENABLE_TPM/fullenc: true
+      ENABLE_SECURE_BOOT/fullenc: true
+      # only secboot does not have a TPM
+      ENABLE_TPM/onlysecboot: false
+      ENABLE_SECURE_BOOT/onlysecboot: true
+      # only tpm does not have secure boot
+      ENABLE_TPM/onlytpm: true
+      ENABLE_SECURE_BOOT/onlytpm: false
+      # no enc does not have secure boot or tpm
+      ENABLE_TPM/noenc: false
+      ENABLE_SECURE_BOOT/noenc: false
```

Comment:
I don't like using the name "secboot" in the variant name, especially in the one that doesn't refer to what we call secboot (measured and secured boot with FDE).

Also 3 out of 4 variants are effectively the same. Because without either secureboot or without tpm or without both, we just do plain boot.

Can't we reduce this to just two variants, always with a tpm, and one variant with secureboot on, second variant with secureboot off?

Or is there something special about testting secureboot-off with tpm or secureboot-on without tpm? I don't think there is any different code paths for those, are there?

Reply:
Sure if you want to eliminate 2 of the variants to just have secure boot w/ TPM (and thus also full disk encryption), as well as the plain boot without any of this that's fine. I included them because there is code in snap-bootstrap that operates slightly differently depending on if we have just secure boot or just a TPM, etc. but we already have tests for this different behavior in snapd, so that setup doesn't need to be replicated here.